From d9641b2fc4986e2a37333a2fa053519767ec23b0 Mon Sep 17 00:00:00 2001
From: Victor Westerhuis <victor@westerhu.is>
Date: Wed, 20 Jul 2022 13:05:35 +0200
Subject: [PATCH 3/4] Make code sent by backspace key configurable

Keep the default code as ASCII backspace, but allow setting the code to
ASCII delete. This is the code used by the Linux console and most modern
terminal emulators by default.

This depends on a corresponding change in libtsm.
---
 meson.build           | 4 +++-
 meson_options.txt     | 4 ++++
 src/kmscon_terminal.c | 3 +++
 src/pty.c             | 9 +++++++--
 4 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/meson.build b/meson.build
index 4544587..650276e 100644
--- a/meson.build
+++ b/meson.build
@@ -45,7 +45,7 @@ systemddir = get_option('libdir') / 'systemd'
 fs = import('fs')
 
 xkbcommon_deps = dependency('xkbcommon', version: '>=0.5.0')
-libtsm_deps = dependency('libtsm', version: '>=4.0.0')
+libtsm_deps = dependency('libtsm', version: '>=4.1.0')
 libudev_deps = dependency('libudev', version: '>=172')
 libdrm_deps = dependency('libdrm')
 dl_deps = dependency('dl')
@@ -117,6 +117,7 @@ endforeach
 config.set('BUILD_ENABLE_DEBUG', get_option('extra_debug'))
 config.set_quoted('BUILD_MODULE_DIR', prefix / moduledir)
 config.set_quoted('BUILD_CONFIG_DIR', prefix / sysconfdir)
+config.set10('BUILD_BACKSPACE_SENDS_DELETE', get_option('backspace_sends_delete'))
 
 # Make all files include "config.h" by default. This shouldn't cause any
 # problems and we cannot forget to include it anymore.
@@ -156,6 +157,7 @@ summary({
   'extra_debug': get_option('extra_debug'),
   'tests': get_option('tests'),
   'docs': enable_docs,
+  'backspace_sends_delete': get_option('backspace_sends_delete'),
 }, section: 'Miscellaneous')
 
 #
diff --git a/meson_options.txt b/meson_options.txt
index 35a10de..7460d5b 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -36,3 +36,7 @@ option('session_dummy', type: 'feature', value: 'auto',
   description: 'dummy session')
 option('session_terminal', type: 'feature', value: 'auto',
   description: 'terminal session')
+
+# terminal options
+option('backspace_sends_delete', type: 'boolean', value: 'false',
+  description: 'backspace sends ASCII delete (0177) instead of ASCII backspace (010)')
diff --git a/src/kmscon_terminal.c b/src/kmscon_terminal.c
index 6edf935..508bd97 100644
--- a/src/kmscon_terminal.c
+++ b/src/kmscon_terminal.c
@@ -624,6 +624,9 @@ int kmscon_terminal_register(struct kmscon_session **out,
 	if (ret)
 		goto err_con;
 
+	tsm_vte_set_backspace_sends_delete(term->vte,
+					   BUILD_BACKSPACE_SENDS_DELETE);
+
 	ret = tsm_vte_set_palette(term->vte, term->conf->palette);
 	if (ret)
 		goto err_vte;
diff --git a/src/pty.c b/src/pty.c
index 1443f4a..3f0f14e 100644
--- a/src/pty.c
+++ b/src/pty.c
@@ -340,8 +340,13 @@ static void setup_child(int master, struct winsize *ws)
 		goto err_out;
 	}
 
-	/* erase character should be normal backspace */
-	attr.c_cc[VERASE] = 010;
+	if (BUILD_BACKSPACE_SENDS_DELETE) {
+		/* erase character should be delete */
+		attr.c_cc[VERASE] = 0177;
+	} else {
+		/* erase character should be normal backspace */
+		attr.c_cc[VERASE] = 010;
+	}
 
 	/* set changed terminal attributes */
 	if (tcsetattr(slave, TCSANOW, &attr) < 0) {
-- 
2.46.0

